* add option for deliver a specific existing build

hjanuschka opened the pull request:
allows deliver to select a specific/existing build number instead of uploading current build binary.


sample:

```ruby
deliver(
	build_number: "666",
	force: true,
	skip_metadata: true,
	skip_screenshots: true,
	submit_for_review: true
)
```

to use the latest  (already uploaded build on Tunes) - set `build_number` to `latest`

KrauseFx commented on lib/deliver/options.rb:
```diff
@@ -90,6 +90,15 @@ def self.available_options
                                      is_string: false,
                                      optional: true),
+        FastlaneCore::ConfigItem.new(key: :build_number,
+                                     description: "if set the given build number will be used instead of the latest build number (combin with skip_upload) ",
```
`combin` should be `combined`, also `If` should be upper case.

KrauseFx commented on lib/deliver/options.rb:
```diff
@@ -90,6 +90,15 @@ def self.available_options
+                                     description: "if set the given build number will be used instead of the latest build number (combin with skip_upload) ",
+                                     is_string: true,
+                                     optional: true),
+        FastlaneCore::ConfigItem.new(key: :skip_binary_upload,
```
We're never using this value anywhere?

KrauseFx commented:
Hey @hjanuschka, thanks for this PR. I really like the idea of submitting existing builds for Review, however I think this change adds quite some complexity. 

How about we design it as following:

```
deliver submit_build --build_number 830 --version_number 1.4
```

while both values are optional. 

This command would then *only* submit the build for review and not upload anything: no metadata, no screenshots and no builds. 

What do you think?

hjanuschka commented:
thx for the review.

 * removed the unused option (has been left from another atempt to solve my problem)
 * i like the idea of having a standalone action (like in match, the nuke actions)
  * the proposal looks nice, without the version/build parameters the latest will be submitted.


but to be honest, i don't think it ads to much complexity - because it is one parameter, and only if set, just selects a specific build instead of waiting for the upload build.

i think letting the user choose to upload screenshots and metadata, is a good way.
combining "select a specific build"  and also update metadata (release notes), screenshots and so on, seems to be a use-case for me.

hjanuschka commented:
ok rebased the PR - the command 

`./deliver submit_build —build_number 199` should no work.

and my variation should also work

hjanuschka commented:
* rebased current master wich has `skip_upload` option

LGTM

lacostej commented on lib/deliver/commands_generator.rb:
```diff
@@ -49,7 +49,31 @@ def run
+          options[:submit_for_review] = true
+          options[:force] = true
 
+          unless loaded
```
this feels a bit dangerous to do here as we pass some options that are today unrelated but could affect the setup one day (e.g. force). No ?

hjanuschka replied:
hmmm, i am not sure - the action states `submit specific build for review` 
should we just set force to `false` and skip the others?

lacostej commented on lib/deliver/submit_for_review.rb:
```diff
@@ -29,10 +29,19 @@ def submit!(options)
       v = app.edit_version
-      build = wait_for_build(app)
+
+      if options[:build_number]
```
you might want to extract `options[:build_number]`

hjanuschka replied:
i am pretty new to ruby, i don't get it, what do you mean by `extract`

lacostej replied:
create a variable :)

KrauseFx replied:
@hjanuschka what he means
```ruby
build_number = options[:build_number]
```
I don't think you need to update this, I'm ok with accessing the values directly here

lacostej commented on lib/deliver/submit_for_review.rb:
```diff
@@ -29,10 +29,19 @@ def submit!(options)
+
+      if options[:build_number]
+        Helper.log.info "Selecting existing build-number: #{options[:build_number]}"
+        build = v.candidate_builds.find(build_version: options[:build_number]).first
```
has this been tested for other values than 'latest' ? I don't think this works. (find takes a block not parameters)

lacostej replied:
![image](https://cloud.githubusercontent.com/assets/24282/12580925/adbacdae-c433-11e5-90fb-d1735e24eff1.png)

:)

hjanuschka commented:
thx @lacostej about the wrong `.find()`  fixed now

hjanuschka commented:
@KrauseFx any chances for a review

KrauseFx commented:
Reviewing this now. I'd also like the confirmation from @jaleksynas, as he wanted this feature :

KrauseFx commented on lib/deliver/commands_generator.rb:
```diff
@@ -49,7 +49,22 @@ def run
           Deliver::Runner.new(options).run
         end
       end
+      command :submit_build do |c|
```
This needs to be added to the README.md too :+1:

KrauseFx commented on lib/deliver/commands_generator.rb:
```diff
@@ -49,7 +49,22 @@ def run
+          options[:skip_metadata] = true
+          options[:skip_screenshots] = true
+          options[:submit_for_review] = true
+          options[:skip_upload] = true
```
Sorry, but I had to rename this option https://github.com/fastlane/deliver/commit/8097b7ab3a87f2e6695d98faeed32b28a1a176cb

KrauseFx commented on lib/deliver/options.rb:
```diff
@@ -94,6 +94,12 @@ def self.available_options
                                      optional: true),
+        FastlaneCore::ConfigItem.new(key: :build_number,
+                                     short_option: "-n",
+                                     description: "If set the given build number will be used instead of the latest build number ",
```
Please remove the space at the end of the string

KrauseFx commented on lib/deliver/runner.rb:
```diff
@@ -20,7 +20,7 @@ def run
       upload_metadata
 
-      has_binary = options[:ipa] || options[:pkg]
+      has_binary = (options[:ipa] || options[:pkg]) && !options[:build_number]
```
Why does `options[:build_number]` have an influence on if we have a binary?

KrauseFx commented on lib/deliver/submit_for_review.rb:
```diff
@@ -29,10 +29,22 @@ def submit!(options)
+      if options[:build_number]
+        Helper.log.info "Selecting existing build-number: #{options[:build_number]}"
+        build = v.candidate_builds.detect { |a| a.build_version == options[:build_number] }
+        unless build
```
We have to show an error if we can't find the build

KrauseFx replied:
Ugh, that seems kind of dirty, we first try to find a build, and then check if it's `"latest"`? How about we check it before and change the first `if` to

```ruby
if options[:build_number] and options[:build_number] != "latest"
```

jaleksynas commented:
I like this feature for certain!  it is critical when using decoupled pilot/testflight operations.  

As a standalone command does this still work inside the fastlane deliver action as shown at the top of the PR? 

My intention is to have a lane that builds, uploads and sends to testing using pilot. (Continuously)

The a user will run a release lane to simply 'deliver' the latest meta-data, screenshots and submit the latest build for review once we bless the latest TestFlight build.  I like the meta-data/screenshots to be part of this deliver step because often this content is being updated during testing with testflight.

hjanuschka commented:
@jaleksynas  like your feedback, thx - going to work on this PR tonight.
the feature works as you described - we have exact the same workflow - and it will work from fastfile and from standalone commandline

hjanuschka commented:
ok feedback implemented

 * rebased current master
 *  added README entry
 * removed all fixed options from command_runner, except the "submit_for_review" the other should work  like on all other deliver calls



@KrauseFx  if no build is found an exception is raised, if `build_number` is not latest.


```
12:30:00]: Activating language de-DE...
[12:30:03]: Uploading metadata to iTunes Connect
[12:30:04]: Successfully uploaded initial set of metadata to iTunes Connect
[12:30:05]: Starting with the upload of screenshots...
[12:30:05]: Successfully uploaded screenshots to iTunes Connect
[12:30:08]: Finished the upload to iTunes Connect
[12:30:09]: Selecting existing build-number: 199
/Users/hja/Desktop/SPORTPORTAL/deliver/lib/deliver/submit_for_review.rb:40:in `select_build': Build number: 199 does not exist (RuntimeError)
	from /Users/hja/Desktop/SPORTPORTAL/deliver/lib/deliver/submit_for_review.rb:5:in `submit!'

```



so for now you can do via commandline:


```
deliver submit_build --build_number 199 --force --skip_metadata # if you want to skip metadata for example
```

via fastfile it will look like this:

```
deliver(
     build_number: 199,
    force: true
)
```


if no skip param is set, metadata, screenshots and so on will be updated, so i hope this fits your workflow @jaleksynas  (we pretty much have the same)


@KrauseFx  - mind if you review again?

KrauseFx commented on README.md:
````diff
@@ -132,6 +132,13 @@ To get a list of available options run
 ```
 
+
+Select an already exisiting build - on the current version train - and submit it for review.
````
Please change to 

```
Select a previously uploaded build and submit it for review
```

KrauseFx commented on lib/deliver/commands_generator.rb:
```diff
@@ -49,7 +49,17 @@ def run
+          options = FastlaneCore::Configuration.create(Deliver::Options.available_options, options.__hash__)
+          options.load_configuration_file("Deliverfile")
+          options[:submit_for_review] = true
+          options[:build_number] = "latest" unless options[:build_number]
```
This looks very much like something that should be the `default_value` of the `: build_number` option :+1:

hjanuschka replied:
well this won't work out - when build_number defaults to a value that is not `false` - it will trigger the submit_build code path - therefore i think it is neccassary to have this portion of ugliness here in the commander part - for keeping the rest of the implementation clean.

KrauseFx replied:
Good point :+1:

KrauseFx commented on lib/deliver/options.rb:
```diff
@@ -94,6 +94,12 @@ def self.available_options
                                      optional: true),
+        FastlaneCore::ConfigItem.new(key: :build_number,
+                                     short_option: "-n",
+                                     description: "If set the given build number  (already uploaded to iTC) will be used instead of the current built one",
```
Double spaces before `(`

KrauseFx commented on lib/deliver/options.rb:
```diff
@@ -94,6 +94,12 @@ def self.available_options
+                                     description: "If set the given build number  (already uploaded to iTC) will be used instead of the current built one",
+                                     default_value: "latest",
+                                     optional: true
+                                    ),
```
Please move to line above

KrauseFx commented on lib/deliver/runner.rb:
```diff
@@ -20,8 +20,8 @@ def run
-      has_binary = options[:ipa] || options[:pkg]
-      if !options[:skip_binary_upload] && has_binary
+      has_binary = (options[:ipa] || options[:pkg])
+      if !options[:skip_binary_upload] && !options[:build_number] && has_binary
```
As I can see here, the user can't provide an `ipa` and a `build_number` (which makes sense), however currently we'll fail silently. 

There is a cool new feature in the fastlane config system you can see here: https://github.com/fastlane/deliver/blob/master/lib/deliver/options.rb#L37-L40

Basically you can define that `skip_binary_upload` doesn't like `ipa` and `pkg`, so that the user will get a nice error message :+1:

KrauseFx commented:
Hey @hjanuschka, this is great, thank you so much for updating the PR, this looks good already. I added some comments - also it seems like there is a merge conflict.

Thank you so much for all your ongoing contribution and help, really appreciated :tada: :rocket: :sparkles: (3 emojis!)

hjanuschka commented:
thx for the review - and for the three emojis!


  * master rebased
  * changed raise and helper.log to UI stuff
  * implemented your suggestions (except the on with the default_value in the commander part, commented on it https://github.com/fastlane/deliver/pull/548#discussion_r52057354)
  * rubocop goot,  tests pass and no hound woof!


 hope all is good and we are ready to merge!

KrauseFx commented:
Perfect, thank you again for updating the PR and working in the last changes. This will be in the next release :+1:
